src/phpDocumentor/Dsn.php

```diff
@@ -0,0 +1,308 @@
+<?php
+/**
+ * This file is part of phpDocumentor.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @copyright 2010-2015 Mike van Riel<mike@phpdoc.org>
+ * @license   http://www.opensource.org/licenses/mit-license.php MIT
+ * @link      http://phpdoc.org
+ */
+
+namespace phpDocumentor;
+
+/**
+ * Value Object for DSN.
+ */
+
+final class Dsn
+{
+    /** @var string */
+    private $scheme;
+
+    /** @var string */
+    private $host;
+
+    /** @var int */
+    private $port;
+
+    /** @var string */
+    private $user;
+
+    /** @var string */
+    private $password;
+
+    /** @var string */
+    private $path;
+
+    /** @var array  */
+    private $query = [];
+
+    /** @var array  */
+    private $parameters = [];
+
+    /** @var array  */
+    private $locationParts = [];
+
+    /** @var array  */
+    private $dsnParts = [];
+
+    /**
+     * @param $dsn
+     */
+    public function __construct($dsn)
+    {
+        $this->parse($dsn);
+    }
+
+    /**
+     * @return string
+     */
+    public function getScheme()
+    {
+        return $this->scheme;
+    }
+
+    /**
+     * @return string
+     */
+    public function getHost()
+    {
+        return $this->host;
+    }
+
+    /**
+     * @return int
+     */
+    public function getPort()
+    {
+        return $this->port;
+    }
+
+    /**
+     * @return string
+     */
+    public function getUsername()
+    {
+        return $this->user;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPassword()
+    {
+        return $this->password;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPath()
+    {
+        return new Path($this->path);
+    }
+
+    /**
+     * @return string
+     */
+    public function getQuery()
+    {
+        return $this->query;
+    }
+
+    /**
+     * @return string
+     */
+    public function getParameters()
+    {
+        return $this->parameters;
+    }
+
+    /**
+     * @param $dsn
+     */
+    private function parse($dsn)
+    {
+        $this->dsnParts = explode(';', $dsn);
+        $location = $this->dsnParts[0];
+        unset($this->dsnParts[0]);
+        $this->locationParts = parse_url($location);
+
+        if (! $this->locationParts) {
+            throw new \InvalidArgumentException(
+                sprintf('"%s" is not a valid DSN.', $dsn)
+            );
+        }
+
+
+        $this->parseScheme();
+
+        $this->parseHostAndPath();
+
+        $this->parsePort();
+
+        $this->parseUser();
+
+        $this->parsePassword();
+
+        $this->parseQuery();
+
+        $this->parseParameters();
+    }
+
+    /**
+     * validates and sets the scheme property
+     */
+    private function parseScheme()
+    {
+        if (array_key_exists('scheme', $this->locationParts)) {
```

Comment:
it is more common to use `isset($this->locationParts['scheme'])` but I don't know if and what the advantage of one over the other is :)